weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -1978,6 +1911,329 @@ def completions_create(
+            chunk_iter, start_call, model_name, req.project_id
+        )
+
+    def _update_metadata_from_chunk(
+        self, chunk: dict[str, Any], aggregated_metadata: dict[str, Any]
+    ) -> None:
+        """Update aggregated metadata from a chunk."""
+        metadata_fields = [
+            "id",
+            "created",
+            "model",
+            "system_fingerprint",
+            "service_tier",
+            "usage",
+        ]
+
+        for field in metadata_fields:
+            if field in chunk and field not in aggregated_metadata:
+                if field == "service_tier":
+                    aggregated_metadata[field] = chunk.get(field, "default")
+                else:
+                    aggregated_metadata[field] = chunk[field]
+
+    def _process_tool_call_delta(
+        self, tool_call_delta: list, tool_calls: list[dict[str, Any]]
+    ) -> None:
+        """Process tool call delta and update tool_calls list."""
+        for tool_call in tool_call_delta:
+            tool_call_index = tool_call.get("index", 0)
+
+            # Ensure we have enough tool calls in our list
+            while len(tool_calls) <= tool_call_index:
+                tool_calls.append(
+                    {
+                        "id": None,
+                        "type": "function",
+                        "function": {"name": "", "arguments": ""},
+                    }
+                )
+
+            existing_tool_call = tool_calls[tool_call_index]
+
+            # Update existing tool call fields
+            if tool_call.get("id"):
+                existing_tool_call["id"] = tool_call["id"]
+            if tool_call.get("type"):
+                existing_tool_call["type"] = tool_call["type"]
+
+            if "function" in tool_call:
+                function_data = tool_call["function"]
+                if function_data.get("name"):
+                    existing_tool_call["function"]["name"] = function_data["name"]
+                if "arguments" in function_data:
+                    existing_tool_call["function"]["arguments"] += function_data[
+                        "arguments"
+                    ]
+
+    def _clean_tool_calls(
+        self, tool_calls: list[dict[str, Any]]
+    ) -> list[dict[str, Any]]:
+        """Clean up tool_calls - remove incomplete ones and ensure proper format."""
+        cleaned_tool_calls = []
+        for tool_call in tool_calls:
+            if tool_call.get("id") and tool_call.get("function", {}).get("name"):
+                cleaned_tool_call = {
+                    "function": {
+                        "arguments": tool_call["function"]["arguments"],
+                        "name": tool_call["function"]["name"],
+                    },
+                    "id": tool_call["id"],
+                    "type": "function",
+                }
+                cleaned_tool_calls.append(cleaned_tool_call)
+        return cleaned_tool_calls
```

Comment:
Nit: use list comprehension. Same logic, slightly more elegant.
```suggestion
        cleaned_tool_calls = [
            {
                "function": {
                    "arguments": tool_call["function"]["arguments"],
                    "name": tool_call["function"]["name"],
                },
                "id": tool_call["id"],
                "type": "function",
            }
            for tool_call in tool_calls
            if tool_call.get("id") is not None and tool_call.get("function", {}).get("name") is not None
        ]

        return cleaned_tool_calls
```